tensorflow_addons/optimizers/weight_decay_optimizers.py

```diff
@@ -226,6 +238,25 @@ def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
                 grad, var, indices, apply_state=apply_state
             )
 
+    def _do_use_weight_decay(self, var):
+        """Whether to use L2 weight decay for `var`."""
+        if not self._decay_var_list or var.ref() in self._decay_var_list:
+            if self.exclude_from_weight_decay:
```

Comment:
Could we maybe factor this part out into a function in optimizer/utils.py as  `exclude_from_weight_decay(variable, exclude_regexes: List[str]) -> bool`? This would ensure consistency between LAMB and potential future optimizers with exclude lists for weight decay and reduce nesting.

Also, I'd consider it to be more intuitive to have decay_var_list having a higher priority over exclude_from_weight decay since it's a function argument (i.e., if decay_var_list is specified, always decay those independent of a match in exclude_weight_decay). WDYT?

Reply:
- I added a function `is_variable_excluded_by_regexes(variable, exclude_regexes: List[str]) -> bool` in `optimizer/utils.py`. As it can also be used in `lamb.py _do_layer_adaptation`. I will change this name if not comfortable for you. It's mostly called as `not is_variable_excluded_by_regexes(...)`, or maybe prefer a function like `is_variable_not_excluded_by_regexes`? WDYT?
- I changed `_do_use_weight_decay` in `optimizers/weight_decay_optimizers.py` that `if self._decay_var_list and var.ref() in self._decay_var_list` will return `True`. Making `decay_var_list` higher priority than `exclude_from_weight_decay`.

Reply:
Thanks for factoring this out and making decay_var_list higher priority!

How about renaming  `is_variable_excluded_by_regexes` to `is_variable_matched_by_regexes(variable, regexes)`? I feel this is a bit easier to parse. I think having the not in front is fine as it keeps the function name simpler.

Reply:
Ya, ranamed `is_variable_excluded_by_regexes` to `is_variable_matched_by_regexes`, also its param `exclude_regexes` to just `regexes`.